api/src/main/java/org/pmiops/workbench/db/dao/WorkspaceServiceImpl.java

```diff
@@ -0,0 +1,112 @@
+   * it here.
+   */
+  @Override
+  public WorkspaceDao getDao() {
+    return workspaceDao;
+  }
+  @Override
+  public void setDao(WorkspaceDao workspaceDao) {
+    this.workspaceDao = workspaceDao;
+  }
+  @Override
+  public Workspace get(String ns, String id) {
+    return workspaceDao.findByWorkspaceNamespaceAndFirecloudName(ns, id);
+  }
+  @Override
+  public Workspace getRequired(String ns, String id) {
+    Workspace workspace = get(ns, id);
+    if (workspace == null) {
+      throw new NotFoundException(String.format("Workspace %s/%s not found.", ns, id));
+    }
+    return workspace;
+  }
+  @Override
+  public List<Workspace> findForReview() {
+    return workspaceDao.findByApprovedIsNullAndReviewRequestedTrueOrderByTimeRequested();
+  }
+
+  // FIXME @Version instead? Bean instantiation v. @Transactional?
+  //@Transactional
+  @Override
+  public void setResearchPurposeApproved(String ns, String id, boolean approved) {
+    Workspace workspace = getRequired(ns, id);
+    if (workspace.getReviewRequested() == null || !workspace.getReviewRequested()) {
+      throw new BadRequestException(String.format(
+          "No review requested for workspace %s/%s.", ns, id));
+    }
+    if (workspace.getApproved() != null) {
+      throw new BadRequestException(String.format(
+          "Workspace %s/%s already %s.",
+          ns, id, workspace.getApproved() ? "approved" : "rejected"));
+    }
+    workspace.setApproved(approved);
+    workspaceDao.save(workspace);
+  }
+
+  @Override
+  @Transactional
+  public void updateUserRoles(String ns, String id, Set<WorkspaceUserRole> userRoleSet) {
+    org.pmiops.workbench.db.model.Workspace dbWorkspace = getRequired(
+        ns, id);
+    Map<Integer, WorkspaceUserRole> userRoleMap = new HashMap<Integer, WorkspaceUserRole>();
+    for (WorkspaceUserRole userRole : userRoleSet) {
+      userRole.setWorkspace(dbWorkspace);
+      int key = Objects.hash(userRole.getUser().getUserId(), userRole.getWorkspace().hashCode());
+      userRoleMap.put(key, userRole);
+    }
+    Iterator<WorkspaceUserRole> dbUserRoles = dbWorkspace.getWorkspaceUserRoles().iterator();
+    while (dbUserRoles.hasNext()) {
+      boolean resolved = false;
+      WorkspaceUserRole currentUserRole = dbUserRoles.next();
+
+      WorkspaceUserRole mapValue = userRoleMap.get(Objects.hash(currentUserRole.getUser().getUserId(),
+          currentUserRole.getWorkspace().hashCode()));
+      if (mapValue != null) {
+        currentUserRole.setRole(mapValue.getRole());
+        userRoleMap.remove(Objects.hash(currentUserRole.getUser().getUserId(),
+            currentUserRole.getWorkspace().hashCode()));
+      } else {
+        dbUserRoles.remove();
+      }
+    }
+
+    for (Map.Entry<Integer, WorkspaceUserRole> remainingRole : userRoleMap.entrySet()) {
+      dbWorkspace.getWorkspaceUserRoles().add(remainingRole.getValue());
+    }
+  }
```

Comment:
So you don't actually have to call save() to make these changes persist?

Reply:
Correct.